### PR TITLE
[WIP] No damage from throwing of inappropriate items

### PR DESCRIPTION
**Pull request by illi-kun**
I tested it with some items and throwing works better with this patch (sorry, no damage from pillows anymore).

As a side note, I want to highlight that throwing of katanas cause _MASSIVE_ damage (OMG, 100+ dmg per hit!), so we need some flag to distinct items like throwing knives and shurikens (which should deal significant cutting damage in case of throwing) from swords (which are almost useless for throwing).

### CODE REVIEW

**Comment by Coolthulhu** [comment]
> we need some flag to distinct items like throwing knives and shurikens

More like a separate statblock, similar to how ranged weapons don't use their ranged stats in melee.

**Comment by Shadefang** [comment]
I wouldn't say that throwing knives and the like should deal a better ratio of damage than larger swords, they should just be easier to throw. Possibly have how much of the cutting damage is applied be based off of the item's size and your skill, possibly with a flag for items designed to be thrown? As for a separate stat block, When wouldn't the damage dealt when thrown be a direct derivative of its melee stats?

**Comment by kevingranade** [comment]
Heh, shuriken shouldn't do much damage, they're for distraction, not
killing.
Probably we should give cutting weapons that are suited to throwing a flag
to mark them as such, and otherwise make application of their cutting
damage very unlikely or impossible.
Bashing weapons should be fine, it's just a matter if hitting, right?

**Comment by kevingranade** [comment]
I like the stat block for particularly good throwing weapons, but we still
need a fallback for improvised orojectiles.

**Comment by Chezzo** [comment]
The lawn dart does no bashing damage, but does 16 cut damage.

**Comment by illi-kun** [comment]
Separate stat block for such thing seems overkill for me because as Shadefang correctly noted general "throwing" cutting property is in direct dependence on general cutting prop.
So, the most reasonable way for me is to deal 100% of cutting dmg to target if item has the discussed flag and max(bashing, cutting * very_small_factor) otherwise.

**Comment by illi-kun** [comment]
> Bashing weapons should be fine, it's just a matter if hitting, right?

Well, if you're able to throw a sledgehammer it should deal solid damage for sure but we don't check that possibility at all and only effect of character's strength is additional amount of damage.

**Comment by illi-kun** [comment]
@Chezzo how it correlates with current code (without this PR)?

**Comment by Coolthulhu** [comment]
> max(bashing, cutting * very_small_factor)

Of what type?
> Separate stat block for such thing seems overkill for me

Consider weapons that currently deal mixed bashing/stabbing, like spears. Spears' bashing damage depends on force behind them and not just their momentum.
Meanwhile a thrown hatchet would have enough momentum to deal some of its bashing damage.

On an unrelated note, is there a particular reason for 0 bashing damage hits still having that 0 damage component rather than not adding it?

File: src/ranged.cpp [hunk 1]
```diff
@@ -247,7 +247,6 @@ dealt_projectile_attack Creature::projectile_attack( const projectile &proj_arg,
             // Don't do anything, especially don't call map::shoot as this would damage the vehicle
         } else {
             g->m.shoot( tp, proj, !no_item_damage && tp == target_arg );
-            has_momentum = proj.impact.total_damage() > 0;
```
Comment (kevingranade) [inline src/ranged.cpp]: I'm going to revert this change, as it doesn't seem to have anything to do with the rest of the change, and I'm pretty sure it's going to break things, in particular it will allow bullets to fly through an unlimited number of walls as long as they can penetrate one.

Comment (kevingranade) [inline src/ranged.cpp]: Nevermind, I see why it's related, but this is still wrong, this code allows the projectile to spend some of its momentum breaking through relatively fragile map features like thin doors and windows and continuing to hit something else, this change means projectiles will keep going through any map features such as walls, even if it is unable to destroy them.
Concrete example, you can throw a grenade through a wall with this line removed.

Comment (illi-kun) [inline src/ranged.cpp]: Thank you for catching this, I suspected that simple removing this line can ruine something important. I'll take a look for some better solution.

**Comment by Zireael07** [comment]
As for katanas dealing massive dmg, how about a cap on the value that gets used, regardless of whether we're using bashing or cutting for the throws?

**Comment by illi-kun** [comment]
@Zireael07 I think _appropriate_ weapon should deal massive dmg elsewise throwing will be useless skill, so a cap is not a proper solution imo.

**Comment by illi-kun** [comment]
Closing this since I don't have better idea how to solve that. I'll create an issue about that instead and probably will try to deal with it later.
